### ISSUE TITLE
react-component needs to be in array

### ISSUE THREAD

**Issue by andrewdeandrade** [mlmorg/react-hyperscript#7]
Currently react-hyperscript requires children react components to be in an array even when there is only one child. Support passing in one child component without having to wrap it in an array.

**Comment by mrozbarry** [comment]
I might be a bit of the niave side of this, but is there any huge performance hit by not enforcing an array at all?

    h('div', {}, firstChild, secondChild, thirdChild);

You can grab all of these children internally using `arguments`, and build an array.

My main cause is that I have been using react with coffeescript, and with this syntax, it would read similar to:

    h 'ol', {},
      h 'li', {}, 'First Item'
      h 'li', {}, 'Second Item'

Internally, you'd have to do something like:

    childArray = Array.prototype.slice.call(arguments, 2);

I think it offers a good compromise.  Backwards compatibility would have to be checked;

    if typeof(arguments[2]) == 'array'
      childArray = arguments[2];
    else
      childArray = Array.prototype.slice.call(arguments, 2);

I don't know if that's a big performance hit, either.  Just a thought.

**Comment by richarddewit** [comment]
@mrozbarry +1

**Comment by fiatjaf** [comment]
@mrozbarry 

```
if typeof(arguments[2]) == 'array'
  childArray = arguments[2];
else
  childArray = Array.prototype.slice.call(arguments, 2);
```

Is a performance hit. And it even doesn't account for cases in which you wouldn't have the second argument as `props`.

**Comment by PascalPixel** [comment]
I have tried hyperscript before, but found that is easier to use `React.createElement()` since it doesn't require an array:

```es6
import React from 'react'
const o = React.createElement

const App = () =>
  o('div', {},
    o('h1', {}, 'Hello'),
    o(Foo),
    'Ok!',
    o('p', {}, 'How are you?'),
    o(Foo, {},
      o(Bar, { key: 'hi' },
        o('span', {}, 'Heya!')
      )
    )
  )
```

It does however, constantly require the `{}` or `null` as a second argument, which is annoying.
At this point, is it correct to assume that dropping the array would mean a security or performance hit?

Ref https://github.com/reactjs/rfcs/issues/45

**Comment by JAForbes** [comment]
I think if it doesn't have a negative impact on performance it'd be good to support this.  Other libraries with similar api's (e.g. mithril) support variadic children.  I often jump between frameworks for different projects and it'd be nice to be able to re-use templates with minimal effort and surprise.

**Comment by JAForbes** [comment]
> found that is easier to use React.createElement() since it doesn't require an array:

True, but it doesn't support css selector syntax, so it's not compatible with some pretty compelling work flows like e.g. [bss](https://github.com/porsager/bss#valueof).

If variadics are supported by `React.createElement`, then that's another reason to support it in `react-hyperscript`.
